src/Microsoft.Kusto.ServiceLayer/DataSource/KustoQueryUtils.cs

```diff
@@ -91,16 +91,22 @@ public static bool  IsClusterLevelQuery(string query)
                 dictionary[key] = new Dictionary<string, T> {{metadata.Name, metadata}};
             }
         }
-        
-        public static void SafeAdd(this Dictionary<string, List<DataSourceObjectMetadata>> dictionary, string key, DataSourceObjectMetadata node)
+
+        public static void SafeAdd<T>(this Dictionary<string, SortedDictionary<string, DataSourceObjectMetadata>> dictionary, string key,
+            T node) where T : DataSourceObjectMetadata
         {
             if (dictionary.ContainsKey(key))
             {
-                dictionary[key].Add(node);
+                if (dictionary[key].ContainsKey(node.PrettyName))
```

Comment:
Minor, but the return statement can be avoided with a negative check:
if(!dictionary[key]....)
{
    dictionary[key].Add(....);
}